### PR TITLE
DTensor.to_local() drops the _is_param marker that nn.Parameter sets on custom-tensor i...

### LINKED ISSUE #166156: Parameter.to_local() will make parameter a normal tensor
### 🐛 Describe the bug

```python
w # a distributed parameter
isinstance(w, nn.Parameter) # True
w = w.to_local()
isinstance(w, nn.Parameter) # False
```

### Versions

PyTorch version: 2.10.0a0+git9166f61
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.19 | packaged by conda-forge | (main, Oct 22 2025, 22:29:10) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-124-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 

Nvidia driver version: 570.124.06
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvtx==0.2.13
[pip3] onnx==1.19.1
[pip3] onnx-ir==0.1.11
[pip3] onnxscript==0.5.4.dev20251017
[pip3] optree==0.17.0
[pip3] torch==2.10.0a0+git9166f61
[pip3] torch_tb_profiler==0.4.3
[pip3] torchao==0.14.0+git02941240f
[pip3] torchdata==0.11.0
[pip3] torchtitan==0.1.0
[pip3] triton==3.4.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by aditvenk**
### Summary
`nn.Parameter(dtensor).to_local()` returned a plain `torch.Tensor` rather than
preserving the `nn.Parameter` identity. After the fix, `isinstance(local, nn.Parameter)`
correctly mirrors `isinstance(dtensor, nn.Parameter)` in both grad-enabled and
no-grad modes.

### Root Cause
`nn.Parameter` represents Parameter-ness for custom tensor subclasses (like
`DTensor`) by setting a `_is_param = True` flag on the instance.
`torch.nn.parameter._ParameterMeta.__instancecheck__` looks at that flag when it
decides whether `isinstance(x, nn.Parameter)` is True.

`DTensor.to_local()` in `torch/distributed/tensor/_api.py` returned either
`self._local_tensor` (no_grad path) or the result of `_ToTorchTensor.apply`
(grad path). Neither carries the `_is_param` flag from the wrapping DTensor, so
the returned tensor was treated as a plain `torch.Tensor`.

### Fix
`torch/distributed/tensor/_api.py`: after computing the local result, if the
DTensor itself is an `nn.Parameter` (has `_is_param`), stamp `_is_param = True`
on the returned tensor. In the no-grad branch the returned tensor IS the
DTensor's internal `_local_tensor`, so we first take a `view_as` (a fresh
tensor object that shares storage) and set the flag on the view — this keeps
the DTensor's internal storage from accidentally becoming a Parameter.

This mirrors the design of `nn.Parameter.__new__` for custom tensors, which
also sets `_is_param` on a fresh instance returned by `detach()`.

We probe `_is_param` via `getattr` rather than `isinstance(self, nn.Parameter)`
because (1) this layer of the stack should not need a `torch.nn` import, and
(2) `_ParameterMeta.__instancecheck__` itself uses the `_is_param` attribute to
decide Parameter-ness for custom tensor subclasses, so the two are equivalent
by construction.

### Repro
The repro script asserts the expected Parameter-ness on the returned local
tensor. Before the fix, the assertions raise and the script exits 1; after the
fix, the assertions pass and the script exits 0.
Run:

```bash
python repro_166156_generated.py
```

<details>
<summary>Repro Script</summary>

```python
"""
Repro for pytorch/pytorch#166156:
Parameter.to_local() will make parameter a normal tensor.

This script asserts that calling .to_local() on an nn.Parameter wrapping a
DTensor preserves its Parameter-ness. It exits non-zero when the bug is
present (before the fix) and exits 0 only when the fix is in place.
"""

import os
import sys
import torch
import torch.nn as nn
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor import distribute_tensor, Replicate


def main() -> int:
    os.environ.setdefault("MASTER_ADDR", "127.0.0.1")
    os.environ.setdefault("MASTER_PORT", "29501")
    os.environ.setdefault("WORLD_SIZE", "1")
    os.environ.setdefault("RANK", "0")
    torch.distributed.init_process_group(backend="gloo")
    try:
        mesh = init_device_mesh("cpu", (1,))
        t = torch.randn(4, 4, requires_grad=True)
        dt = distribute_tensor(t, mesh, [Replicate()])
        w = nn.Parameter(dt)

        print(
            "Before to_local: "
            f"type={type(w).__name__}, "
            f"isinstance Parameter: {isinstance(w, nn.Parameter)}"
        )
        # Sanity check: the wrapped DTensor IS a Parameter at this point.
        assert isinstance(w, nn.Parameter), (
            "Precondition failed: nn.Parameter(DTensor) should itself be an "
            "nn.Parameter."
        )

        local = w.to_local()
        print(
            "After to_local (grad enabled): "
            f"type={type(local).__name__}, "
            f"isinstance Parameter: {isinstance(local, nn.Parameter)}"
        )

        with torch.no_grad():
            local_nograd = w.to_local()
        print(
            "After to_local (no grad): "
            f"type={type(local_nograd).__name__}, "
            f"isinstance Parameter: {isinstance(local_nograd, nn.Parameter)}"
        )

        # These assertions encode the reported failure: before the fix in
        # torch/distributed/tensor/_api.py:DTensor.to_local, both of these
        # return a plain Tensor, not an nn.Parameter, and the assertions fail
        # (script exits non-zero). After the fix, both pass.
        assert isinstance(local, nn.Parameter), (
            "BUG #166156: nn.Parameter(DTensor).to_local() returned a "
            "non-Parameter tensor in grad-enabled mode."
        )
        assert isinstance(local_nograd, nn.Parameter), (
            "BUG #166156: nn.Parameter(DTensor).to_local() returned a "
            "non-Parameter tensor in no_grad mode."
        )

        print("OK: nn.Parameter-ness preserved by DTensor.to_local()")
        return 0
    finally:
        torch.distributed.destroy_process_group()


if __name__ == "__main__":
    sys.exit(main())
```

</details>

### Testing
`test_fix.py` (extended edge cases) all pass:

```
Replicate Parameter preserved: OK
Shard Parameter preserved: OK
Parameter(requires_grad=False) preserved: OK
Non-Parameter DTensor stays plain tensor: OK
Internal _local_tensor not mutated: OK
Gradient flow through to_local: OK

All checks passed.
```

A new regression test was added at
`test/distributed/tensor/test_dtensor.py::DTensorTest::test_to_local_preserves_parameter`.
The DTensor test harness defaults to NCCL on a host with CUDA but no NCCL
build (this sandbox), so the in-tree distributed test suite could not be run
in-place here. Logic of the test exactly mirrors the standalone checks above,
which all pass; the in-tree test should be exercised in CI before merge.

The in-tree regression test covers the autograd-related concerns flagged by
the reviewers:

- Asserts `local.requires_grad` is True for a grad-requiring Parameter.
- Calls `.sum().backward()` on the returned local Parameter and asserts
  `dtensor_param.grad is not None`, confirming autograd still flows back into
  the DTensor parameter after the `_is_param` marker is stamped on the
  non-leaf result of `_ToTorchTensor.apply`.
- Asserts the internal `_local_tensor` of the DTensor is NOT mutated into a
  Parameter in either path (the `view_as` guard for the no_grad branch).
- Negative case: a plain (non-Parameter) DTensor must NOT become a Parameter
  after `to_local()`.

### Existing assertion preserved
`test_dtensor_constructor` (same file) asserts that in the no-grad case the
returned local tensor IS `sharded_tensor._local_tensor`. That test wraps a raw
`DTensor` rather than an `nn.Parameter`, so the new `_is_param` branch does not
trigger and the identity-equals behavior remains unchanged.

### Lint
`spin fixlint torch/distributed/tensor/_api.py test/distributed/tensor/test_dtensor.py`
reports "No lint issues" on both lintrunner passes (RUFF/PYREFLY/... and
PYFMT/CODESPELL/FLAKE8/...).

## Files Changed
- `torch/distributed/tensor/_api.py`
- `test/distributed/tensor/test_dtensor.py`

Fixes #166156

---
*This PR was generated by [ptq](https://github.com/drisspg/pt_job_queue) with human review.*

### CODE REVIEW

File: torch/distributed/tensor/_api.py [hunk 1]
```diff
@@ -602,13 +602,33 @@ def to_local(
...
-            return self._local_tensor
+            result = self._local_tensor
+        else:
+            if grad_placements is not None and not isinstance(grad_placements, tuple):
+                grad_placements = tuple(grad_placements)
+            result = _ToTorchTensor.apply(
+                self, grad_placements
+            )  # pyre-ignore[16]: autograd func
+
+        # Preserve nn.Parameter-ness: if self is an nn.Parameter (i.e. has the
```
Comment (aditvenk) [inline torch/distributed/tensor/_api.py:613]: Could this break backward compatibility (BC)?

Comment (aditvenk) [reply torch/distributed/tensor/_api.py:613]: [bot] Resolved in the latest update: Addressed in the diff: torch/distributed/tensor/_api.py:DTensor.to_local now carries an inline 'BC:' paragraph stating that the new code path only runs when self has _is_param set (i.e. self is already a Parameter(DTensor)), so every raw-DTensor call site is byte-for-byte unchanged, including the historical `to_local() is self._local_tensor` identity guarantee in the no-grad branch which is covered by the existing test_dtensor_constructor test. The PR description / report.md also contains a 'Backward compatibility analysis' section enumerating the five BC scenarios (raw DTensor identity preserved, type(result) unchanged, nn.Parameter re-wrap round-trip, optimizer step over Parameter(DTensor), DTensor._local_tensor not mutated) all verified by test_bc.py. The single narrow observable change is for Parameter(DTensor) in no_grad mode where `param.to_local() is param._local_tensor` is no...

**Comment by aditvenk** [comment]
@claude Review and check if failing tests on CI are caused by this change

**Review by sanketpurandare — APPROVED** [review]

**Comment by aditvenk** [comment]
@claude

**Comment by aditvenk** [comment]
@claude In the no-grad case, could there be a problem if we are constructing the view_as within no-grad but the user later mutates the tensor with grad-enabled? Could this be a regression in capability compared to before?

**Comment by aditvenk** [comment]
@claude any of the failing tests in CI caused by this change?
